File: 102_binary_tree_level_order_traversal.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        traversal = []
        queue = []
        cur_level = 0
        if root:
            queue.append((0, root))
        
        while len(queue) > 0:
            level = []
            while len(queue) > 0 and queue[0][0] == cur_level:
                cur_node = queue.pop(0)[1]
                level.append(cur_node.val)
                if cur_node.left:
                    queue.append((cur_level + 1, cur_node.left))
                if cur_node.right:
                    queue.append((cur_level + 1, cur_node.right))
            traversal.append(level)
            cur_level += 1
        
        return traversal

File: c++/102_binary_tree_level_order_traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> traversal;
        int visit_depth = -1;
        queue<pair<int, TreeNode*>> visit;
        if (root != nullptr) {
            visit.push(make_pair(0, root));
        }

        while (!visit.empty()) {
            // Pop from front of queue
            int cur_depth;
            TreeNode* cur;
            tie(cur_depth, cur)= visit.front();
            visit.pop();

            // Add current node val to appropriate traversal level
            if (cur_depth == visit_depth) {
                traversal[visit_depth].push_back(cur->val);
            } else {
                traversal.push_back({cur->val});
                visit_depth = cur_depth;
            }

            // Enqueue any existant children
            if (cur->left != nullptr) {
                visit.push(make_pair(cur_depth + 1, cur->left));
            }
            if (cur->right != nullptr) {
                visit.push(make_pair(cur_depth + 1, cur->right));
            }
        }

        return traversal;
    }
};
